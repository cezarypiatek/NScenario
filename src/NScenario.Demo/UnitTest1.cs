using System;
using System.Threading.Tasks;
using NUnit.Framework;

namespace NScenario.Demo
{
    public class Tests
    {

        [Test]
        public async Task should_present_basic_scenario()
        {
            var scenario = TestScenarioFactory.Default();

            await scenario.Step("This is the first step", () =>
            {
                // Here comes the logic
            });
            
            await scenario.Step("This is the second step", () =>
            {
                // Here comes the logic
            });
            
            await scenario.Step("This is the third step", () =>
            {
                // Here comes the logic
            });
        }
        
        

        [Test] 
        public async Task shouldPresentBASICScenarioWithCamelCaseTitle()
        {
            var scenario = TestScenarioFactory.Default();

            await scenario.Step("This is the first step", () =>
            {
                // Here comes the logic
            });
            
            await scenario.Step("This is the second step", () =>
            {
                // Here comes the logic
            });
            
            await scenario.Step("This is the third step", () =>
            {
                // Here comes the logic
            });
        }
        
        [TestCase(false)]
        [TestCase(true)]
        public async Task should_present_basic_scenario_with_explicit_title(bool someFlag)
        {
            var scenario = TestScenarioFactory.Default(title: $"some scenario when flag set to '{someFlag}'");

            await scenario.Step("This is the first step", () =>
            {
                // Here comes the logic
            });
            
            await scenario.Step("This is the second step", () =>
            {
                // Here comes the logic
            });
            
            await scenario.Step("This is the third step", () =>
            {
                // Here comes the logic
            });
        }
        
        [Test]
        public async Task should_present_basic_scenario_with_steps_returning_value()
        {
            var scenario = TestScenarioFactory.Default();

            var valFromStep1 = await scenario.Step("This is the first step", () =>
            {
                // Here comes the logic
                return 1;
            });

            var valFromStep2 = await scenario.Step("This is the second step", async () =>
            {
                // Here comes the logic
                await Task.Yield();
                return valFromStep1 + 1;
            });

            await scenario.Step("This is the third step", () =>
            {
                // Here comes the logic
                _ = valFromStep1 + valFromStep2;
            });
        }
        
        [Test]
        public async Task should_present_scenario_with_sub_steps()
        {
            var scenario = TestScenarioFactory.Default();

            await scenario.Step("This is the first step", async () =>
            {
                await scenario.Step("This is the first sub-step of first step", () =>
                {
                    // Here comes the logic
                });
                await scenario.Step("This is the second sub-step of first step", async () =>
                {
                    await scenario.Step("Yet another nesting level p1", () =>
                    {
                        // Here comes the logic
                    });
                    await scenario.Step("Yet another nesting level p2", () =>
                    {
                        // Here comes the logic
                    });
                });
            });
            
            await PerformReusableScenarioPart(scenario);
            
            await scenario.Step("This is the third step", () =>
            {
                // Here comes the logic
            });
        }

        private static async Task PerformReusableScenarioPart(ITestScenario scenario)
        {
            await scenario.Step("This is the second step", async () =>
            {
                await scenario.Step("This is the first sub-step of second step", () =>
                {
                    // Here comes the logic
                });
                await scenario.Step("This is the second sub-step of second step", () =>
                {
                    // Here comes the logic
                });
            });
        }
    }
}