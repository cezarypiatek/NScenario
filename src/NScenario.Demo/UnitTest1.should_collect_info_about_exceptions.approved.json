{
  "ScenarioTitle": "should collect info about exceptions",
  "MethodName": "should_collect_info_about_exceptions",
  "FilePath": "_IGNORED_VALUE_",
  "LineNumber": 145,
  "Status": 1,
  "Steps": [
    {
      "Description": "This is the first step",
      "LineNumber": 149,
      "FilePath": "_IGNORED_VALUE_",
      "ExecutionTime": "_IGNORED_VALUE_",
      "Status": 2,
      "Exception": null,
      "SubSteps": [
        {
          "Description": "This is the first sub-step of first step",
          "LineNumber": 151,
          "FilePath": "_IGNORED_VALUE_",
          "ExecutionTime": "_IGNORED_VALUE_",
          "Status": 3,
          "Exception": null,
          "SubSteps": null
        },
        {
          "Description": "This is the second sub-step of first step",
          "LineNumber": 155,
          "FilePath": "_IGNORED_VALUE_",
          "ExecutionTime": "_IGNORED_VALUE_",
          "Status": 2,
          "Exception": null,
          "SubSteps": [
            {
              "Description": "Yet another nesting level p1",
              "LineNumber": 157,
              "FilePath": "_IGNORED_VALUE_",
              "ExecutionTime": "_IGNORED_VALUE_",
              "Status": 2,
              "Exception": "System.InvalidOperationException: Something wrong\r\n   at NScenario.Demo.Tests.<>c.<should_collect_info_about_exceptions>b__5_4() in C:\\repos\\NScenario\\src\\NScenario.Demo\\UnitTest1.cs:line 159\r\n   at NScenario.StepExecutors.OutputScenarioStepExecutor.Step(String scenarioName, String stepDescription, Func`1 action, StepContext stepContext) in C:\\repos\\NScenario\\src\\NScenario\\StepExecutors\\OutputScenarioStepExecutor.cs:line 31\r\n   at NScenario.StepExecutors.LevelTrackingScenarioStepExecutor.Step(String scenarioName, String stepDescription, Func`1 action, StepContext stepContext) in C:\\repos\\NScenario\\src\\NScenario\\StepExecutors\\LevelTrackingScenarioStepExecutor.cs:line 21\r\n   at NScenario.StepExecutors.LevelTrackingScenarioStepExecutor.Step(String scenarioName, String stepDescription, Func`1 action, StepContext stepContext) in C:\\repos\\NScenario\\src\\NScenario\\StepExecutors\\LevelTrackingScenarioStepExecutor.cs:line 21\r\n   at NScenario.ScenarioInfoCollectorExecutor.Step(String scenarioName, String stepDescription, MaybeAsyncAction action, StepContext stepContext) in C:\\repos\\NScenario\\src\\NScenario\\ScenarioInfoCollectorExecutor.cs:line 61",
              "SubSteps": null
            }
          ]
        }
      ]
    }
  ]
}